osquery/tables/system/windows/ntfs_acl_permissions.cpp

```diff
@@ -0,0 +1,227 @@
+            TLOG << "LookupAccountSid error: " << GetLastError();
+		}
+		else {
+			return name;
+		}
+	case TRUSTEE_IS_NAME:
+		// get the name from ptstrName
+		return t.ptstrName;
+	case TRUSTEE_BAD_FORM:
+		// invalid
+		return "Invalid";
+	case TRUSTEE_IS_OBJECTS_AND_SID:
+		// ptstrName member is a pointer to an OBJECTS_AND_SID struct
+        if (!LookupAccountSid(
+				NULL,
+                (PSID)((OBJECTS_AND_SID*)t.ptstrName)->pSid,
+                name,
+                &size,
+                domain,
+                &size,
+                &accountType)) {
+			//printf("LookupAccountSid error: %u\n", GetLastError());
+            TLOG << "LookupAccountSid error: " << GetLastError();
+		}
+		else {
+			return name;
+		}
+	case TRUSTEE_IS_OBJECTS_AND_NAME:
+		// ptstrName member is a pointer to an OBJECTS_AND_NAME struct
+		return ((OBJECTS_AND_NAME_*) t.ptstrName)->ptstrName;
+	default:
+		return "";
+	}
+}
+
+QueryData genNTFSACLPerms(QueryContext& context) {
+  QueryData results;
+
+	std::string ObjName;
+	unsigned long result = 0;
+	PACL DACL = NULL;
+	PEXPLICIT_ACCESS aceList = NULL, aceItem;
+	unsigned long aceCount, aceIndex;
+
+	auto paths = context.constraints["path"].getAll(EQUALS);
+    for (const auto& path_string : paths) {
+          if (fs::exists(path_string)) {
+				fs::path ObjName = path_string;
+
+				// Get a pointer to the existing DACL.
+
+                result = GetNamedSecurityInfo(path_string.c_str(),
+                                                SE_FILE_OBJECT,
+                                                DACL_SECURITY_INFORMATION,
+                                                NULL,
+                                                NULL,
+                                                &DACL,
+                                                NULL,
+                                                NULL);
+                if (ERROR_SUCCESS != result) {
+                    //printf("GetNamedSecurityInfo Error %u\n", result);
+					TLOG << "GetExplicitEnteriesFromAcl Error " << result;
+                }
+
+				// get list of ACEs from DACL pointer
+                result = GetExplicitEntriesFromAcl(
+                    DACL, &aceCount, &aceList);
+                if (ERROR_SUCCESS != result) {
+                    //printf("GetExplicitEnteriesFromAcl Error %u\n",
+                    //        result);
+					TLOG << "GetExplicitEnteriesFromAcl Error " << result;
+                }
+
+                // Loop through list of entries
+
+                aceItem = aceList;
+                // printf("AccessMode | TrusteeName | AccessPermissions
+                // | Inheritence\n");
+                for (aceIndex = 0; aceIndex < aceCount; aceItem++) {
+                    Row r;
+                    std::string accessMode;
+                    std::string inheritedFrom;
+                    std::string perms;
+                    std::string trusteeName;
```

Comment:
It would be much better to crate all of them in place (probably less memory allocations). Like:
```c++
auto  perms = accessPermsToStr(aceItem->grfAccessPermissions);
auto  accessMode = accessModeToStr(aceItem->grfAccessMode);
auto  inheritedFrom = inheritanceToStr(aceItem->grfInheritance);
auto trusteeName = trusteeToStr(aceItem->Trustee);
```